_posts/2024-08-14-running-a-hsm-realm.md

```diff
@@ -0,0 +1,221 @@
+through this blog post.
+
+## Hardware
+
+The HSMs are PCIe cards and need physical servers to be installed in. The
+services running on these servers use various services from [Google Cloud
+Platform](https://console.cloud.google.com/). Ideally your servers are located
+somewhere that is close to a GCP zone. You'll need the following hardware in
+order to initialize and run a realm.
+
+- 3 or more [Entrust nShield Solo XC HSM's](https://www.entrust.com/products/hsm/nshield-solo). The base model is fine.
+- A [CodeSafe](https://www.entrust.com/products/hsm/codesafe) license for each HSM.
+- A Pack of Entrust blank smart cards.
+- x64 servers of your choice to host the HSM PCIe cards.
+- x64 servers to run the load balancer and cluster manager services.
+  Alternatively these 2 services can be run in a virtualized environment that is
+  located in the same facility as the HSM servers.
+
+## Realm Cluster Sizing
+
+The realm should include at least 2 cluster manager instances and 2 load
+balancer instances. Larger realms may need more load balancers depending on the
+particular specifications of the servers used and the number of HSMs in the
+realm.
+
+Local testing can be done with a single HSM, however test clusters should have a
+minimum of 3 HSMs in order to realistically exercise the consensus and failover
+operations. Production clusters should have at least 5 HSMs, as once the key
+ceremony has been performed additional HSMs can not be added to the realm. A 5
+HSM cluster gives some runway to handle failures.
+
+In our testing a 5 HSM cluster (with 3 replication groups) should be able to
+perform approximately 240 recover operations a second. This throughput drops off
+slightly as the size of the merkle tree increases. We observed a 5% drop in
+throughput going from a tree with 10M leaves to a tree with 500M leaves. Scaling
+to larger throughput will require more HSMs and replication groups.
+
+Improvements to overall availability can be achieved by running multiple
+independent realms with a N of M client configuration. e.g. 2 HSM realms and 1
+software realm with a 2 of 3 client configuration.
+
+## Google Cloud Platform
+
+The GCP services Bigtable, Secret Manager and Cloud Pub/Sub are used by the
+realm services. Credentials to access these will be needed for all hosts
+(regardless of role) in the cluster. Bigtable must be configured for a single
+zone as multi-zone configurations change the [semantics of write
+operations](https://cloud.google.com/bigtable/docs/replication-overview#consistency-model).
+
+## HSM Initialization
+
+For test clusters the `entrust_ops` tool that is part of the [HSM
+repo](https://github.com/juicebox-systems/juicebox-hsm-realm) can be used to
+initialize the NVRAM, create the Security World and the relevant realm keys. The
+[`entrust_hsm/README.md`](https://github.com/juicebox-systems/juicebox-hsm-realm/blob/main/entrust_hsm/README.md)
```

Comment:
We already reference this readme earlier and tell them to keep it available for reference, so we probably don't need to link it again? Also I wonder if it would be helpful to surface some of the `entrust_ops` commands here directly with like step-by-step sort of instructions do get an HSM initialized? Even if it's just what's in the start of that readme. It would make the mentioning of need for blank smart cards above a bit more meaningful too since then they actually clearly get used in this process

Reply:
I'll see if I reword this. if we put the steps in here it'll end up being a copy of about half of the entrust readme which seems a bit extreme.